src/wasm-type-shape.h

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_wasm_type_shape_h
+#define wasm_wasm_type_shape_h
+
+#include <functional>
+#include <vector>
+
+#include "wasm-type.h"
+
+namespace wasm {
+
+// Provides hashing and equality comparison for a sequence of types. The hashing
+// and equality differentiate the top-level structure of each type in the
+// sequence and the equality of referenced heap types that are not in the
+// recursion group, but for references to types that are in the recursion group,
+// it considers only the index of the referenced type within the group. That
```

Comment:
If it considers the index in the group, wouldn't these end up different?
```wat
(rec
  (type $A (sub (struct (field (ref $B)))))
  (type $B (sub (struct)))
)
(rec
  (type $B (sub (struct)))
  (type $A (sub (struct (field (ref $B)))))
)
```

Reply:
Yes indeed, although those rec groups are already different just by looking at their top-level shapes. It would consider these different as well, where the top-level shapes are all the same:

```
(rec
  (type $A (struct (field (ref $A))))
  (type $B (struct (field (ref $B))))
)
(rec
  (type $A (struct (field (ref $A))))
  (type $B (struct (field (ref $A))))
)
```

This precisely matches the spec's notion of rec group identity.

Reply:
Oh, thanks, that's what I was missing. So this is matching the spec behavior, and not defining another notion. I thought a broader notion might make sense when looking at permutations of rec groups but I may have misunderstood the connection of this PR to your other work.

Reply:
Yep, this PR is providing the tools for detecting when the spec would consider two rec groups to the be same so we can then take action to differentiate them.